cpp/src/arrow/json/chunked-builder.cc

```diff
@@ -0,0 +1,401 @@
+      std::shared_ptr<Array> converted;
+      RETURN_NOT_OK(converter_->Convert(unconverted, &converted));
+      std::unique_lock<std::mutex> lock(mutex_);
+      chunks_[block_index] = std::move(converted);
+      return Status::OK();
+    });
+  }
+};
+
+class InferringChunkedArrayBuilder : public NonNestedChunkedArrayBuilder {
+ public:
+  InferringChunkedArrayBuilder(const std::shared_ptr<internal::TaskGroup>& task_group,
+                               const PromotionGraph* promotion_graph,
+                               std::shared_ptr<Converter> converter)
+      : NonNestedChunkedArrayBuilder(task_group, std::move(converter)),
+        promotion_graph_(promotion_graph) {}
+
+  void Insert(int64_t block_index, const std::shared_ptr<Field>& unconverted_field,
+              const std::shared_ptr<Array>& unconverted) override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (chunks_.size() <= static_cast<size_t>(block_index)) {
+      chunks_.resize(static_cast<size_t>(block_index) + 1, nullptr);
+      unconverted_.resize(chunks_.size(), nullptr);
+    }
+    unconverted_[block_index] = unconverted;
+    lock.unlock();
+    ScheduleConvertChunk(block_index, unconverted_field);
+  }
+
+  void ScheduleConvertChunk(int64_t block_index,
+                            const std::shared_ptr<Field>& unconverted_field) {
+    task_group_->Append([this, block_index, unconverted_field] {
+      return TryConvertChunk(static_cast<size_t>(block_index), unconverted_field);
+    });
+  }
+
+  Status TryConvertChunk(size_t block_index,
+                         const std::shared_ptr<Field>& unconverted_field) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    auto converter = converter_;
+    auto unconverted = unconverted_[block_index];
+    std::shared_ptr<Array> converted;
+
+    lock.unlock();
+    Status st = converter_->Convert(unconverted, &converted);
+    lock.lock();
+
+    if (converter != converter_) {
+      // another task promoted converter; reconvert
+      lock.unlock();
+      ScheduleConvertChunk(block_index, unconverted_field);
+      return Status::OK();
+    }
+
+    if (st.ok()) {
+      // conversion succeeded
+      chunks_[block_index] = std::move(converted);
+      unconverted_[block_index] = unconverted;
+      return Status::OK();
+    }
+
+    auto promoted_type =
+        promotion_graph_->Promote(converter_->out_type(), unconverted_field);
+    if (promoted_type == nullptr) {
+      // converter failed, no promotion available
+      return st;
+    }
+    RETURN_NOT_OK(MakeConverter(promoted_type, converter_->pool(), &converter_));
+
+    size_t nchunks = chunks_.size();
+    for (size_t i = 0; i < nchunks; ++i) {
+      if (i != block_index && chunks_[i]) {
+        // We're assuming the chunk was converted using the wrong type
+        // (which should be true unless the executor reorders tasks)
+        chunks_[i].reset();
+        lock.unlock();
+        ScheduleConvertChunk(i, unconverted_field);
```

Comment:
Unconverted_field is used in several places to store JSON kind of an unexpected field (since both string and number fields are string arrays until conversion). I could definitely store JSON kind alongside `unconverted_[i]` so that the field isn't being passed around so much.